Enable int8xint8->int32 `einsum`

james77777778 opened the pull request:
jax:
use `preferred_element_type="int32"` to enable int8xint8->int32 in `einsum`

tf:
add custom ops for 2 subscripts below to replace `tf.einsum` with `tf.matmul` in order to enable int8xint8->int32 format:
- "abcde,afce->acdbf"
- "abc,dce->abde"

other backends:
just align the dtype conversion with jax and tf.

However, I'm not sure how to benchmark them, and currently, I don't see much improvement by using int8 `einsum` compared to float32/bfloat16 `einsum` in jax and tf backends.

fchollet commented:
Thanks for the PR!

> However, I'm not sure how to benchmark them, and currently, I don't see much improvement by using int8 einsum compared to float32/bfloat16 einsum in jax and tf backends.

You can simply create a `@jax.jit` or `@tf.function(jit_compile=True)` function that calls einsum, and call it repeatedly it for large random arrays with the 2 supported subscripts. Large would be something in line with array sizes in LLMs, e.g. `(32, 512, 512)` for the input. Make sure to not count the initial compilation time (so only count after the 1st execution).

What numbers are you currently seeing for different backends?

james77777778 commented:
@fchollet 
Thank you for the reasonable settings!

I've adjusted the benchmark script from #19137  to ensure that the computation will utilize nearly 100% of my 4070.

```python
import time

from keras import backend
from keras import ops
from keras import random

SUBSCRIPTS = "abc,dce->abde"
# [b, dim, dim], [dim, dim, dim] -> [b, dim, dim, dim]


def int8_linear_layer(x, params):
    # read in x and w as pre-quantized int8 tensors
    y = ops.einsum(SUBSCRIPTS, x, params["w"])

    # add bias and apply activation in fp32
    y = ops.cast(y, "float32")
    y = y + params["b"]
    y = ops.relu(y)

    # quantize and store output as int8
    y = ops.round(y / params["s"])
    y = ops.clip(y, -128, 127)
    y = ops.cast(y, "int8")
    y = y[..., 0]
    return y


def fp_linear_layer(x, params):
    y = ops.relu(ops.add(ops.einsum(SUBSCRIPTS, x, params["w"]), params["b"]))
    y = y[..., 0]
    return y


def linear_layer(x, params):
    if backend.standardize_dtype(params["w"].dtype) == "int8":
        return int8_linear_layer(x, params)
    else:
        return fp_linear_layer(x, params)


def network(x, network_params):
    for layer_params in network_params:
        x = linear_layer(x, layer_params)
    return x[0, 0]


def benchmark(batch_size, dim, dtype, n_layers=32, iters=200):
    """Benchmarks the throughput of a linear layer.

    matmul + bias + act (+ quantization)

    To drown out data transfer costs and other overheads, each run performs a
    forward pass of a network consisting of `n_layers` of linear layers in
    series.

    The full network is run for `iters` iterations to compute the throughput in
    TFLOPS/TOPS.
    """

    # create input tensor for network
    if dtype == "int8":
        x = random.uniform(
            [batch_size, dim, dim], minval=-127, maxval=127, dtype="float32"
        )
        x = ops.cast(x, "int8")
    else:
        x = random.uniform([batch_size, dim, dim], dtype=dtype)

    # create weight + bias (+ quant scale) tensors for each layer
    network_params = []
    for i in range(n_layers):
        if dtype == "int8":
            w = random.uniform(
                [dim, dim, dim], minval=-127, maxval=127, dtype="float32"
            )
            w = ops.cast(w, "int8")
            b = random.normal([dim], dtype="float32")
            s = random.normal([], dtype="float32")
            layer_params = {"w": w, "b": b, "s": s}
        else:
            w = random.uniform([dim, dim, dim], dtype=dtype)
            b = random.uniform([dim], dtype=dtype)
            layer_params = {"w": w, "b": b}
        network_params.append(layer_params)

    # jit
    if backend.backend() == "tensorflow":
        import tensorflow as tf

        jit_network = tf.function(network, jit_compile=True)
    elif backend.backend() == "jax":
        import jax

        jit_network = jax.jit(network)
    elif backend.backend() == "torch":
        import torch

        torch.set_float32_matmul_precision("high")
        jit_network = torch.compile(network)

    # warmup
    for _ in range(10):
        _ = ops.convert_to_numpy(jit_network(x, network_params))

    times = []
    for _ in range(iters):
        t0 = time.time()
        _ = ops.convert_to_numpy(jit_network(x, network_params))
        elapsed_time = time.time() - t0  # in ms
        times.append(elapsed_time)

    avg_time = sum(times) / len(times)
    gflops = (n_layers * 2 * dim**3) / avg_time * 1e-9
    return gflops


if __name__ == "__main__":
    batch_size = 16

    # not used, just triggers tensorflow/xla initial log dump
    benchmark(batch_size, 64, "float16")

    print("\nGOPS/GFLOPS")
    cols = f"{'dim':<6} | {'int8':>8} {'fp16':>8} {'fp32':>8}"
    print("=" * len(cols) + "\n" + cols + "\n" + "-" * len(cols))
    for dim in [32, 64, 128, 256]:
        int8_time = benchmark(batch_size, dim, "int8", 16)
        fp16_time = benchmark(batch_size, dim, "float16", 16)
        fp32_time = benchmark(batch_size, dim, "float32", 16)
        print(
            f"{dim:<6} | {int8_time:>8.2f} {fp16_time:>8.2f} {fp32_time:>8.2f}"
        )
    print("=" * len(cols))

```

And the results:
```bash
# tensorflow
GOPS/GFLOPS
===================================
dim    |     int8     fp16     fp32
-----------------------------------
32     |     1.82     2.10     1.93
64     |     9.98     9.20     6.56
128    |    10.62     9.15     4.14
256    |    11.71     8.04     3.97
===================================

# jax
GOPS/GFLOPS
===================================
dim    |     int8     fp16     fp32
-----------------------------------
32     |     6.47     6.07     4.81
64     |    19.31    14.84     8.56
128    |    11.18     9.46     4.19
256    |    11.85     8.10     3.98
===================================
```

Note that the performance of int8 is not as efficient as the results of `matmul` but it still the best one.

james77777778 commented:
Do we need more subscripts for tensorflow?
I'm not sure if the 2 included ones are sufficient.

fchollet commented:
Nice -- these numbers are in the right ballpark!

> Do we need more subscripts for tensorflow?
> I'm not sure if the 2 included ones are sufficient.

Can you check BERT, GPT-2, Mistral and Gemma from KerasNLP? Just instantiate the model, grab a MHA layer from the decoder blocker (it is usually a custom layer with "attention" in the name), and print `self.equation` for the `query_dense`, `value_dense`, `key_dense` and `output_dense` layers. They may be named e.g. `_query_dense`/etc.

If we cover the equations used by these layers in these models, then we are good!

In many cases it's possible to figure out the equations by looking at the code, without instantiating the model, e.g. https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/models/gemma/gemma_attention.py

- "btd,ndh->btnh"
- "bsd,kdh->bskh"
- "bsd,kdh->bskh"
- "btnh,nhd->btd"

james77777778 commented:
> Can you check BERT, GPT-2, Mistral and Gemma from KerasNLP? Just instantiate the model, grab a MHA layer from the decoder blocker (it is usually a custom layer with "attention" in the name), and print `self.equation` for the `query_dense`, `value_dense`, `key_dense` and `output_dense` layers. They may be named e.g. `_query_dense`/etc.

I have written a script to get all possible equations from all KerasNLP models:
Note: some equations are hard-coded in KerasNLP
https://github.com/search?q=repo%3Akeras-team%2Fkeras-nlp%20ops.einsum&type=code

<details>

```python
import functools
import string

import keras_nlp
from keras_nlp.src.models.bloom.bloom_backbone import BloomBackbone
from keras_nlp.src.models.bloom.bloom_decoder import BloomDecoder
from keras_nlp.src.models.deberta_v3.disentangled_attention_encoder import (
    DisentangledAttentionEncoder,
)
from keras_nlp.src.models.electra.electra_backbone import ElectraBackbone
from keras_nlp.src.models.gemma.gemma_decoder_block import GemmaDecoderBlock
from keras_nlp.src.models.llama.llama_backbone import LlamaBackbone
from keras_nlp.src.models.llama.llama_decoder import LlamaDecoder
from keras_nlp.src.models.mistral.mistral_transformer_decoder import (
    MistralTransformerDecoder,
)
from keras_nlp.src.models.t5.t5_backbone import T5Backbone
from keras_nlp.src.models.whisper.whisper_backbone import WhisperBackbone
from keras_nlp.src.models.whisper.whisper_decoder import WhisperDecoder
from keras_nlp.src.models.whisper.whisper_encoder import WhisperEncoder


@functools.lru_cache(512)
def _normalize_einsum_subscripts(subscripts):
    # string.ascii_letters
    mapping = {}
    normalized_subscripts = ""
    for c in subscripts:
        if c in string.ascii_letters:
            if c not in mapping:
                mapping[c] = string.ascii_letters[len(mapping)]
            normalized_subscripts += mapping[c]
        else:
            normalized_subscripts += c
    return normalized_subscripts


model_configs = [
    (
        keras_nlp.models.AlbertBackbone.from_preset,
        {"preset": "albert_base_en_uncased", "load_weights": False},
    ),
    (
        keras_nlp.models.BartBackbone.from_preset,
        {"preset": "bart_base_en", "load_weights": False},
    ),
    (
        keras_nlp.models.BertBackbone.from_preset,
        {"preset": "bert_tiny_en_uncased", "load_weights": False},
    ),
    (
        BloomBackbone.from_preset,
        {"preset": "bloom_560m_multi", "load_weights": False},
    ),
    (
        keras_nlp.models.DebertaV3Backbone.from_preset,
        {"preset": "deberta_v3_extra_small_en", "load_weights": False},
    ),
    (
        keras_nlp.models.DistilBertBackbone.from_preset,
        {"preset": "distil_bert_base_en_uncased", "load_weights": False},
    ),
    (
        ElectraBackbone,
        {
            "vocab_size": 10,
            "num_layers": 2,
            "num_heads": 2,
            "hidden_dim": 2,
            "embedding_dim": 2,
            "intermediate_dim": 4,
            "max_sequence_length": 5,
        },
    ),
    (
        keras_nlp.models.GemmaBackbone.from_preset,
        {"preset": "gemma_2b_en", "load_weights": False},
    ),
    (
        keras_nlp.models.GPT2Backbone.from_preset,
        {"preset": "gpt2_base_en", "load_weights": False},
    ),
    (
        keras_nlp.models.FNetBackbone.from_preset,
        {"preset": "f_net_base_en", "load_weights": False},
    ),
    (
        LlamaBackbone,
        {
            "vocabulary_size": 10,
            "num_layers": 2,
            "num_query_heads": 4,
            "num_key_value_heads": 2,
            "hidden_dim": 8,
            "intermediate_dim": 8,
            "max_sequence_length": 10,
        },
    ),
    (
        keras_nlp.models.MistralBackbone.from_preset,
        {"preset": "mistral_7b_en", "load_weights": False},
    ),
    (
        keras_nlp.models.OPTBackbone.from_preset,
        {"preset": "opt_125m_en", "load_weights": False},
    ),
    (
        keras_nlp.models.RobertaBackbone.from_preset,
        {"preset": "roberta_base_en", "load_weights": False},
    ),
    (
        T5Backbone.from_preset,
        {"preset": "t5_small_multi", "load_weights": False},
    ),
    (
        keras_nlp.models.XLMRobertaBackbone.from_preset,
        {"preset": "xlm_roberta_base_multi", "load_weights": False},
    ),
    (
        WhisperBackbone.from_preset,
        {"preset": "whisper_tiny_en", "load_weights": False},
    ),
]

all_possible_equations = set()

for config in model_configs:
    if "preset" in config[1]:
        print(f"Model: {config[1]['preset']}")
    else:
        print(f"Model: {config[0].__name__}")
    model = config[0](**config[1])
    for layer in model.layers:
        print(f"  {layer.name}")
        if isinstance(
            layer,
            (
                keras_nlp.layers.TransformerEncoder,
                keras_nlp.layers.TransformerDecoder,
                DisentangledAttentionEncoder,
                MistralTransformerDecoder,
                LlamaDecoder,
                WhisperDecoder,
                WhisperEncoder,
            ),
        ):
            for eq in [
                layer._self_attention_layer._query_dense.equation,
                layer._self_attention_layer._key_dense.equation,
                layer._self_attention_layer._value_dense.equation,
                layer._self_attention_layer._output_dense.equation,
            ]:
                all_possible_equations.add(_normalize_einsum_subscripts(eq))
            if hasattr(layer._self_attention_layer, "_dot_product_equation"):
                eq = layer._self_attention_layer._dot_product_equation
                all_possible_equations.add(_normalize_einsum_subscripts(eq))
            if hasattr(layer._self_attention_layer, "_combine_equation"):
                eq = layer._self_attention_layer._combine_equation
                all_possible_equations.add(_normalize_einsum_subscripts(eq))
            if isinstance(
                layer, (keras_nlp.layers.TransformerDecoder, WhisperDecoder)
            ):
                if layer._cross_attention_layer is None:
                    continue
                for eq in [
                    layer._cross_attention_layer._query_dense.equation,
                    layer._cross_attention_layer._key_dense.equation,
                    layer._cross_attention_layer._value_dense.equation,
                    layer._cross_attention_layer._output_dense.equation,
                ]:
                    all_possible_equations.add(_normalize_einsum_subscripts(eq))
                if hasattr(
                    layer._cross_attention_layer, "_dot_product_equation"
                ):
                    eq = layer._cross_attention_layer._dot_product_equation
                    all_possible_equations.add(_normalize_einsum_subscripts(eq))
                if hasattr(layer._cross_attention_layer, "_combine_equation"):
                    eq = layer._cross_attention_layer._combine_equation
                    all_possible_equations.add(_normalize_einsum_subscripts(eq))
        elif isinstance(layer, BloomDecoder):
            for eq in [
                layer._self_attention_layer._query_dense.equation,
                layer._self_attention_layer._key_dense.equation,
                layer._self_attention_layer._value_dense.equation,
            ]:
                all_possible_equations.add(_normalize_einsum_subscripts(eq))
        elif isinstance(layer, GemmaDecoderBlock):
            for eq in [
                layer.attention.query_dense.equation,
                layer.attention.key_dense.equation,
                layer.attention.value_dense.equation,
                layer.attention.output_dense.equation,
                layer.gating_ffw.equation,
                layer.gating_ffw_2.equation,
                layer.ffw_linear.equation,
            ]:
                all_possible_equations.add(_normalize_einsum_subscripts(eq))


# Addtional equations from the codes of KerasNLP
for eq in [
    "btkgh,bskh->bkgts",
    "bkgts,bskh->btkgh",
    "i,j->ij",
    "bnqd,bnkd->bnqk",
    "aecd,abcd->acbe",
    "acbe,aecd->abcd",
]:
    all_possible_equations.add(_normalize_einsum_subscripts(eq))

print(f"All possible equations: {all_possible_equations}")

```

</details>

```bash
All possible equations: {'abcd,adbe->acbe', 'a,b->ab', 'abcde,aebf->adbcf', 'abcde,afce->acdbf', 'abcd,abed->abce', 'abc,cde->abde', 'abcd,cde->abe', 'abc,dce->abde', 'abcd,aecd->aceb', 'abc,cd->abd', 'abcd,aecd->acbe'}
```

I will try to include all the equations in this PR. Just need some time 😅

TODO:
- [ ] 'abcd,adbe->acbe'
- [ ] 'a,b->ab'
- [ ] 'abcde,aebf->adbcf'
- [x] 'abcde,afce->acdbf'
- [ ] 'abcd,abed->abce'
- [ ] 'abc,cde->abde'
- [ ] 'abcd,cde->abe'
- [x] 'abc,dce->abde'
- [ ] 'abcd,aecd->aceb'
- [ ] 'abc,cd->abd'
- [ ] 'abcd,aecd->acbe'

fchollet commented:
Wow, that's actually a lot more than I expected.

fchollet reviewed (approved):
LGTM -- the code looks good! If the feature is ready, then we can merge.

james77777778 commented:
> LGTM -- the code looks good! If the feature is ready, then we can merge.

I have verified that there is no performance regression after the changes (all done in `"float32"`):

```bash
# master branch
subscripts=a,b->ab           (dim=128), cost time=0.10466ms(±0.00062)
subscripts=abc,cd->abd       (dim=128), cost time=0.11970ms(±0.00055)
subscripts=abc,cde->abde     (dim=128), cost time=0.12020ms(±0.00050)
subscripts=abc,dce->abde     (dim=128), cost time=1.35215ms(±1.30167)
subscripts=abcd,abed->abce   (dim=64 ), cost time=0.14666ms(±0.00067)
subscripts=abcd,adbe->acbe   (dim=64 ), cost time=0.78262ms(±0.37625)
subscripts=abcd,aecd->acbe   (dim=64 ), cost time=0.79861ms(±0.37621)
subscripts=abcd,aecd->aceb   (dim=64 ), cost time=0.79831ms(±0.37650)
subscripts=abcd,cde->abe     (dim=128), cost time=1.33974ms(±1.29205)
subscripts=abcde,aebf->adbcf (dim=32 ), cost time=0.93679ms(±0.44432)
subscripts=abcde,afce->acdbf (dim=32 ), cost time=1.74974ms(±0.17599)
```

```bash
# this pr
subscripts=a,b->ab           (dim=128), cost time=0.10347ms(±0.00054)
subscripts=abc,cd->abd       (dim=128), cost time=0.11793ms(±0.00059)
subscripts=abc,cde->abde     (dim=128), cost time=0.11889ms(±0.00060)
subscripts=abc,dce->abde     (dim=128), cost time=1.36989ms(±1.32058)
subscripts=abcd,abed->abce   (dim=64 ), cost time=0.11818ms(±0.00048)
subscripts=abcd,adbe->acbe   (dim=64 ), cost time=0.79939ms(±0.37633)
subscripts=abcd,aecd->acbe   (dim=64 ), cost time=0.79967ms(±0.37628)
subscripts=abcd,aecd->aceb   (dim=64 ), cost time=0.79946ms(±0.37636)
subscripts=abcd,cde->abe     (dim=128), cost time=1.34044ms(±1.29371)
subscripts=abcde,aebf->adbcf (dim=32 ), cost time=0.93809ms(±0.44360)
subscripts=abcde,afce->acdbf (dim=32 ), cost time=0.95274ms(±0.44201)
```

In `subscripts="abcde,afce->acdbf"`, the custom implementation get even faster performance than pure `tf.einsum`.

The benchmark script:
<details>

```python
import time

import tensorflow as tf

from keras import ops


def get_inputs(subscripts, dim=64):
    x1_labels = subscripts.split(",")[0]
    x2_labels = subscripts.split("->")[0][len(x1_labels) + 1 :]
    x1_shape = [dim] * len(x1_labels)
    x2_shape = [dim] * len(x2_labels)
    x1 = ops.ones(x1_shape)
    x2 = ops.ones(x2_shape)
    return x1, x2


def benchmark(subscripts, iterations=1000):
    if len(subscripts) >= len("abcde,afce->acdbf"):
        dim = 32
    elif len(subscripts) >= len("abcd,abed->abce"):
        dim = 64
    else:
        dim = 128
    x1, x2 = get_inputs(subscripts, dim=dim)

    @tf.function(jit_compile=True)
    def fn(x1, x2):
        return ops.einsum(subscripts, x1, x2)

    # warmup
    for _ in range(20):
        _ = fn(x1, x2)
    cost_times = []
    for _ in range(iterations):
        t0 = time.time()
        _ = fn(x1, x2)
        cost_times.append(time.time() - t0)
    cost_times = ops.convert_to_tensor(cost_times)
    cost_times = ops.sort(cost_times)
    cost_times = cost_times[int(iterations * 0.25) : -int(iterations * 0.25)]
    mean = ops.convert_to_numpy(ops.mean(cost_times)) * 1000.0
    std = ops.convert_to_numpy(ops.std(cost_times)) * 1000.0
    print(
        f"subscripts={subscripts:<17} (dim={dim:<3}), "
        f"cost time={mean:.5f}ms(±{std:.5f})"
    )


for subscripts in [
    "a,b->ab",
    "abc,cd->abd",
    "abc,cde->abde",
    "abc,dce->abde",
    "abcd,abed->abce",
    "abcd,adbe->acbe",
    "abcd,aecd->acbe",
    "abcd,aecd->aceb",
    "abcd,cde->abe",
    "abcde,aebf->adbcf",
    "abcde,afce->acdbf",
]:
    benchmark(subscripts)

```

</details>

I think this PR is ready now.
@fchollet

fchollet commented:
Awesome work! 🚀
